### PR TITLE
Fix mixed-dtype gradient norms and clipping

### PR DESCRIPTION
**Pull request by aroshanghias-nvd**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Group gradients by dtype before fused L2-norm and clipping-scale calls, so mixed BF16/FP32 lists produce the intended norm and in-place scaling.

The fused backend tested here interprets a list using its first tensor's dtype. For example, `[BF16[128], FP32[128]]` filled with `1/64` returns `0.2165063471` instead of `0.25`. Reversing the order can read beyond the logical BF16 view. Scaling has an independent defect: even with coefficient `0.5` supplied directly, mixed lists leave some elements unscaled or modify padding beyond the gradient view. The reproductions use padded backing allocations.

The change:

- Buckets L2-norm inputs locally and sums their squared norms before the existing distributed reductions. This preserves collective order for empty ranks and ranks with different dtype sets, and accumulates into the existing zero-valued CUDA tensor.
- Buckets both scalar- and tensor-coefficient scaling calls, and avoids calling fused kernels with empty lists.
- Adds focused numerical regressions for both mixed orders, decoupled gradients, scalar- and tensor-coefficient clipping, a non-vector-aligned length, backing-storage preservation, empty gradients, and asymmetric rank ownership. Ordinary `.grad` is also exercised by the mixed-dtype determinism tests.
- Adds mixed-dtype norm/clipping bit-exact replay under side-stream contention to the existing registered optimizer determinism tests.

The existing single-device gradient-list contract is retained. No kernel implementation or optimizer-state layout changes are required.

## Validation

The GPU component results below describe the original fix at `333ecc9`. Validation of the review cleanup is recorded separately below.

- Native two-rank GB200 investigation with PyTorch `2.11.0+cu130` and the Transformer Engine backend: all 48 norm boundary cases per rank matched first-dtype byte interpretation, with 15 incorrect logical norms. Both scaling APIs failed on both mixed orders. Bucketed controls preserved padding and matched the logical reference.
- Matched caller regression: the original had 24 numerical failures in 96 cases per rank; the corrected caller passed those cases plus four empty-list cases per rank. Mixed-buffer and FP32-buffer DDP/DistributedOptimizer/FusedAdam fixtures each completed eight arms per rank.
- The production file on the PR base is byte-identical to the original file used for that investigation. The PR adds typing and repository formatting to the validated implementation.
- Repository `tools/autoformat.sh`: Black, isort, pylint, and Ruff passed. Mypy could not resolve Torch/TE/Apex/pytest imports in the isolated local lint environment.
- `python3 tools/check_kernel_determinism_coverage.py --base-ref origin/main`: passed.
- Exact PR tests on two GB200 ranks with the native TE backend: the original caller produced 29 expected failures out of 105 unit cases per rank; the fixed caller passed all 105 with no skips or errors. Both mixed-dtype determinism cases passed on both ranks, including four bit-exact replays with side-stream contention. Test dependencies were staged separately from the pinned runtime; unrelated top-level suite fixtures were excluded for this focused run.

Review cleanup (`09f6e03`): DFW job `19072876` completed on two H100 80GB GPUs with PyTorch `2.11.0+cu130` and Transformer Engine `2.15.0+42b84005`. All 17 unit cases and both mixed-dtype determinism cases passed on each rank, with no failures, errors, or skips. Each determinism case includes four bit-exact replays with side-stream contention. The runner verified exact source hashes and native TE dispatch; unrelated top-level suite fixtures were excluded. Local formatting/lint and the determinism coverage check also pass.

The reporter in [#7338](https://github.com/NVIDIA/Megatron-LM/issues/7338#issuecomment-5725704816) also completed a three-step H100 training smoke after adapting the caller patch to their older `dev` checkout. This supports the correction in that configuration, rather than validating a complete checkout of this PR's current head.

This establishes the component defect and correction; it does not claim a measured full-training quality impact or that the fused API promises heterogeneous-list support.

## Issue tracking

Related issue: #7338.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added bit-exact determinism tests under the existing registry entry
- [x] I have added proper typing to my code
- [ ] I have added relevant documentation
- [x] I have run `tools/autoformat.sh` on my PR

No training recipe or public API changes; no functional-test goldens or user documentation updates are included.

### CODE REVIEW

**Comment by Connor-XY** [comment]
Linking the community report in #7338: it describes the same mixed BF16/FP32 norm/clipping caller issue and now includes a single-GPU reproducer plus an H100 training case. This PR appears to cover the proposed caller-side correction, including both clipping-scale paths.

@aroshanghias-nvd, please link that report in the PR description and let us know what remains before this is ready for review. The reporter's configuration would provide useful validation alongside the component tests already described here.

**Comment by aroshanghias-nvd** [comment]
Author review is complete, and I’ve linked #7338. The component and determinism tests pass as described. Validation of commit `333ecc9` in the reporter’s H100 training configuration remains outstanding and would complement those tests. I’m marking this ready for review.

**Comment by Connor-XY** [comment]
@wujingyue, could you review the mixed-dtype norm and clipping changes now that this is ready? The reporter has added a successful three-step H100 training smoke in https://github.com/NVIDIA/Megatron-LM/issues/7338#issuecomment-5725704816. They adapted the caller patch from 333ecc9 to their older dev checkout (43e45e13), so this supports the fix in their configuration rather than validating the complete PR checkout at its current head.

File: megatron/core/optimizer/clip_grads.py [hunk 1]
```diff
@@ -117,12 +125,15 @@ def get_grad_norm_fp32(
...
-            )
-            # Since we will be summing across data parallel groups,
-            # we need the pow(norm-type).
-            total_norm = grad_norm**norm_type
+            for index, dtype_grads in enumerate(_group_grads_by_dtype(grads_for_norm)):
+                grad_norm, _ = multi_tensor_applier(
+                    l2_norm_impl, dummy_overflow_buf, [dtype_grads], False
+                )
+                # Combine local squared norms before the existing collectives.
+                if index == 0:
```
Comment (wujingyue) [inline megatron/core/optimizer/clip_grads.py]: Can we simply initialize total_norm to 0.0 and get rid of this condition?

File: megatron/core/optimizer/clip_grads.py [hunk 2]
```diff
@@ -117,12 +125,15 @@ def get_grad_norm_fp32(
             # Use apex's multi-tensor applier for efficiency reasons.
             # Multi-tensor applier takes a function and a list of list
             # and performs the operation on that list all in one kernel.
-            grad_norm, _ = multi_tensor_applier(
-                l2_norm_impl, dummy_overflow_buf, [grads_for_norm], False  # no per-parameter norm
-            )
-            # Since we will be summing across data parallel groups,
-            # we need the pow(norm-type).
-            total_norm = grad_norm**norm_type
+            for index, dtype_grads in enumerate(_group_grads_by_dtype(grads_for_norm)):
```
Comment (wujingyue) [inline megatron/core/optimizer/clip_grads.py]: Naming: dtype_grads is confusing. Consider group_grads or grads.

File: megatron/core/optimizer/clip_grads.py [hunk 3]
```diff
@@ -117,12 +125,15 @@ def get_grad_norm_fp32(
...
+                grad_norm, _ = multi_tensor_applier(
+                    l2_norm_impl, dummy_overflow_buf, [dtype_grads], False
```
Comment (wujingyue) [inline megatron/core/optimizer/clip_grads.py]: It's hard for readers to figure out what `False` mean. Consider a kwarg.

File: megatron/core/optimizer/clip_grads.py [hunk 4]
```diff
@@ -187,13 +198,18 @@ def clip_grad_by_total_norm_fp32(
         assert (
             multi_tensor_scale_tensor_impl is not None
         ), "clip_coeff is tensor type. But multi_tensor_scale_tensor not available."
-        multi_tensor_applier(
-            multi_tensor_scale_tensor_impl, dummy_overflow_buf, [grads, grads], clip_coeff
-        )
+        for dtype_grads in _group_grads_by_dtype(grads):
```
Comment (wujingyue) [inline megatron/core/optimizer/clip_grads.py]: ditto

File: tests/unit_tests/determinism/kernels/manifest.py [hunk 5]
```diff
@@ -419,7 +419,8 @@ class KernelEntry:
         ),
         tests=(K + "test_optimizer_kernels.py",),
         kind="external-lib",
-        notes="multi_tensor l2norm / scale (TE, apex or local fallback) and fused Adam; "
+        notes="multi_tensor l2norm / scale (TE, apex or local fallback), including mixed-dtype "
+        "caller bucketing and padded-storage replay, and fused Adam; "
```
Comment (wujingyue) [inline tests/unit_tests/determinism/kernels/manifest.py]: Why this change?

File: tests/unit_tests/optimizer/test_clip_grads.py
Comment (wujingyue) [inline tests/unit_tests/optimizer/test_clip_grads.py:1]: The test looks pretty verbose. For coverage, I'd just test one config that triggers mixed grad_dtype.

File: tests/unit_tests/optimizer/test_clip_grads.py [hunk 6]
```diff
@@ -59,3 +61,117 @@ def test_no_gtp_pad_zeros_attribute_counts_all_zeros(self):
...
+        clip_grads.clip_grad_by_total_norm_fp32(
+            params, max_norm, total_norm, use_decoupled_grad=use_decoupled_grad
+        )
+
+        for actual, reference in zip(bases, expected):
+            torch.testing.assert_close(actual, reference, rtol=0, atol=0)
+
+    @pytest.mark.parametrize('use_decoupled_grad', [False, True])
+    @pytest.mark.parametrize('tensor_coefficient', [False, True])
+    def test_empty_gradients_do_not_launch_kernel(
```
Comment (wujingyue) [inline tests/unit_tests/optimizer/test_clip_grads.py]: This is arguably testing implementation not behavior. It makes sense to verify grad clipping doesn't crash with empty gradients, but whether a kernel gets launched behind the scene is an implementation detail that I'd rather not test.

**Review by wujingyue — APPROVED** [review]
https://google.github.io/eng-practices/review/reviewer/speed.html#lgtm-with-comments

**Comment by wujingyue** [comment]
Again, this PR should go to @NVIDIA/mcore-optimizer. I happened to touch the optimizer code recently for FSDP, but I don't own the code.

File: megatron/core/optimizer/clip_grads.py (hunk 1 above)
Comment (aroshanghias-nvd) [reply megatron/core/optimizer/clip_grads.py]: Simplified to accumulate into the existing zero-valued CUDA tensor and removed the first-group conditional. Keeping the tensor initialization, rather than a Python float, also handles ranks with no local gradients when they reach the existing reductions. The reductions remain outside the grouping loop.

File: megatron/core/optimizer/clip_grads.py (hunk 2 above)
Comment (aroshanghias-nvd) [reply megatron/core/optimizer/clip_grads.py]: Renamed to `group_grads` in the norm loop and both clipping loops.

File: megatron/core/optimizer/clip_grads.py (hunk 4 above)
Comment (aroshanghias-nvd) [reply megatron/core/optimizer/clip_grads.py]: Renamed to `group_grads` in the norm loop and both clipping loops.

File: megatron/core/optimizer/clip_grads.py (hunk 3 above)
Comment (aroshanghias-nvd) [reply megatron/core/optimizer/clip_grads.py]: Added `# no per-tensor norms` to explain the flag. I kept it positional because the supported local `multi_tensor_applier` wrapper accepts `*args`, not keyword arguments (`megatron/core/utils.py`).

File: tests/unit_tests/determinism/kernels/manifest.py (hunk 5 above)
Comment (aroshanghias-nvd) [reply tests/unit_tests/determinism/kernels/manifest.py]: The wording was intended to document the added mixed-dtype replay coverage. The existing optimizer entry already points to that test file, so I removed the redundant manifest change. The coverage checker still passes with the new replay tests.

File: tests/unit_tests/optimizer/test_clip_grads.py
Comment (aroshanghias-nvd) [reply tests/unit_tests/optimizer/test_clip_grads.py:1]: Reduced the new numerical test matrix from 102 cases to 14, using one gradient length and the precision-aware decoupled-gradient path. I retained both dtype orders because they exercise different byte-width failures, and both coefficient APIs because norm correctness alone does not guarantee correct scaling. The small norm layout matrix retains empty and disjoint rank ownership coverage. Ordinary `.grad` remains exercised by the mixed-dtype determinism tests.

File: tests/unit_tests/optimizer/test_clip_grads.py (hunk 6 above)
Comment (aroshanghias-nvd) [reply tests/unit_tests/optimizer/test_clip_grads.py]: Agreed. Replaced the kernel-launch assertion with a direct empty-gradient clipping call and CUDA synchronization, so the test checks successful completion rather than how the operation is implemented.

File: tests/unit_tests/optimizer/test_clip_grads.py [hunk 7]
```diff
@@ -59,3 +61,82 @@ def test_no_gtp_pad_zeros_attribute_counts_all_zeros(self):
         )
 
         assert num_zeros == grad.numel() - 1
+
+
+_DTYPES = [(torch.bfloat16, torch.float32), (torch.float32, torch.bfloat16)]
```
Comment (wujingyue) [inline tests/unit_tests/optimizer/test_clip_grads.py:66]: This is hard to read. The name doesn't say **why** the dtypes are in tuples. 

Also, it's on the boundary of https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html. Consider letting each test manages its own parametrization with a well-documented intention.

File: tests/unit_tests/optimizer/test_clip_grads.py [hunk 8]
```diff
@@ -59,3 +61,82 @@ def test_no_gtp_pad_zeros_attribute_counts_all_zeros(self):
...
+        squared_norm = torch.zeros(1, device='cuda', dtype=torch.float64)
+        for grad in grads:
+            squared_norm += grad.double().square().sum()
+        torch.distributed.all_reduce(squared_norm)
+
+        actual = clip_grads.get_grad_norm_fp32(
+            grads, grad_stats_parallel_group=torch.distributed.group.WORLD
+        )
+
+        assert float(actual) == pytest.approx(float(squared_norm.sqrt()), rel=2e-6, abs=1e-8)
```
Comment (wujingyue) [inline tests/unit_tests/optimizer/test_clip_grads.py:109]: torch.testing.assert_close instead of reinventing another way to compare?

File: tests/unit_tests/optimizer/test_clip_grads.py [hunk 9]
```diff
@@ -59,3 +61,82 @@ def test_no_gtp_pad_zeros_attribute_counts_all_zeros(self):
...
+        clip_grads.clip_grad_by_total_norm_fp32(
+            params, 1.0000005, total_norm, use_decoupled_grad=True
+        )
+
+        for actual, reference in zip(bases, expected):
+            torch.testing.assert_close(actual, reference, rtol=0, atol=0)
+
+    @pytest.mark.parametrize('tensor_coefficient', [False, True])
+    def test_clip_empty_gradients(self, tensor_coefficient):
+        if tensor_coefficient and clip_grads.multi_tensor_scale_tensor_impl is None:
```
Comment (wujingyue) [inline tests/unit_tests/optimizer/test_clip_grads.py:138]:
```suggestion
```
IIUC, CI installs TE so this check is not so relevant. It's like we could ask every test to `skipif` `not torch.cuda.is_available` but nobody really cares :)

**Review by wujingyue — APPROVED** [review]

File: megatron/core/optimizer/clip_grads.py (hunk 3 above)
Comment (wujingyue) [reply megatron/core/optimizer/clip_grads.py]: > multi_tensor_applier wrapper accepts *args, not keyword arguments (megatron/core/utils.py).

Got it -- that's a bummer but worth a separate PR :)
